proposals/optional-iteration.md

```diff
@@ -0,0 +1,85 @@
+# Optional Iteration
+
+* Proposal: [SE-NNNN](NNNN-filename.md)
+* Authors: [Anthony Latsis](https://github.com/AnthonyLatsis)
+* Review Manager: TBD
+* Status: **Awaiting implementation**
+
+## Introduction
+
+Optionals are a key feature of Swift and a powerful tool that seamlessly interacts with code;
+they comprise an elegant and concise syntax that serves a great means in expressing
+"act accordingly if there's a value, skip otherwise".
+Some vivid examples are optional chaining, optional invocation `foo?()`, `if let`, `guard let` and `for let`. This proposal considers further supporting this convenience in `for-in` loops.
+
+Swift-evolution thread: [Discussion thread topic for that proposal](https://forums.swift.org/t/another-try-at-allowing-optional-iteration/14376?u=anthonylatsis)
+
+## Motivation
+
+Loops are a common statement in almost every codebase. When working with optional sequences in Swift, a possibility to optionally iterate
+(that is, iterate if there is a value, otherwise skip) is self-explanatory. However, Swift currently doesn't offer a way to directly express this, in the language of optionals: optional sequences are illegal as a `for-in` loop attribute. The most common way of accomplishing this is
+
+```swift
+if let sequence = optionalSequence {
+  for element in sequence { ... }
+}
+```
+
+To avoid additional nesting, you can coalesce with `?? []` for `Array` or use `sequence?.forEach`, which excludes usage of control transfer statements.
+
+## Proposed solution
+
+This proposal introduces optional iteration (`for?`) and hence the possibility to use optional sequences as the sequence attribute in `for-in` loops. 
+
+``` swift 
+var array: [Int]? = nil
+
+for? element in array { ... } // Does nothing
+
+array = [1, 2, 3]
+
+for? element in array { ... }
+// Equivalent to
+for element in array! { ... }
```

Comment:
I would clarify what the syntactic construct is equivalent to in the general case:

```
if let array2 = array {
  for element in array2 {
    ...
  }
}
```

This summary suggests that the force-unwrap is part of the behavior of `for?`, which might mislead someone reading this bit in isolation.